### PR TITLE
Delete ports

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

The PR starts to attempt to correctly model the goose client
abstractions in a way that allows for individual goose clients to have
their own http.Client.

With this step it now makes it possible for nova and neutron to have
their own http.Headers. Previously it wasn't possible to tell nova to
have one set of headers and for neutron to have it's own.

It was changed so we can override the neutron http.Headers independently
from nova. When only affecting neutron it means our potential risk is
lower as we're not touching nova REST APIs.

---

I've split out the client related stuff into a client.go file. This was
to help grok the changes, but if it makes it easier to review I'm more
than happy to put them back and move them in another commit.

I also introduced a ClientFactory which hopefully makes it much easier
where the clients come from.

---

Also added a few comments on the way through and a few TODOs.

## QA steps

Bootstrap onto every OpenStack available.

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -468,7 +468,8 @@
 
 [[constraint]]
   name = "gopkg.in/goose.v2"
-  revision = "ab914003d547f0dc01c26ac8421c55c6b1d66ace"
+  revision = "97fbb7d298bfd7dd9d5c80d697171c54045c1b5c"
+  source = "github.com/SimonRichardson/goose"
```
Comment (SimonRichardson) [inline Gopkg.toml:472]: This will be removed once we like this PR.

File: provider/openstack/provider.go [hunk 2]
```diff
@@ -2124,6 +2063,14 @@ func (e *Environ) terminateInstances(ctx context.ProviderCallContext, ids []inst
 				break
 			}
 		}
+
+		// Attempt to destroy the ports that could have been created when using
+		// spaces.
+		if err := e.terminateInstanceNetworkPorts(id); err != nil {
```
Comment (manadart) [inline provider/openstack/provider.go:2069]: If we destroy the server first, the ports will no longer be associated with it.
We need to either delete them before, or collect the port IDs before.

File: provider/openstack/client_test.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package openstack
+
+import (
+	"github.com/golang/mock/gomock"
+	"github.com/juju/juju/cloud"
```
Comment (manadart) [inline provider/openstack/client_test.go:8]: Grouping.

**Comment by SimonRichardson** [comment]
!!build!!

> FAIL: caasoperator_test.go:429: WorkerSuite.TestContainerStart

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**